### ISSUE TITLE
Error: Running the contributed command:'extension.run0' failed.

### ISSUE THREAD

**Issue by GerjanOnline** [GuardRex/vscode-status-bar-tasks#1]
I am running into this error, any idea? 

![image](https://cloud.githubusercontent.com/assets/1153419/13233176/bbfd8a0c-d9b2-11e5-9699-3fe720e41b50.png)

![image](https://cloud.githubusercontent.com/assets/1153419/13233179/c96d0ece-d9b2-11e5-90cf-fcbeb1e2bc40.png)

**Comment by guardrex** [comment]
@GerjanOnline Yeah ... it actually uses the `args` of each task in order to know what command to run. Therefore, your `tasks.json` tasks must have this format (note the `args` in the task) ...

``` json
{
    "taskName": "restore",
    "suppressTaskName": true,
    "args" : ["dotnet", "restore"],
    "showOutput": "always",
    "problemMatcher": "$msCompile"
},
```

Can you try this format for your file and see if it will work ...

``` json
{
    "version": "0.1.0",
    "command": "npm"
    "isShellCommand": true,
    "showOutput": "always",
    "args": ["run"],
    "tasks": [
        {
            "taskName": "build",
            "suppressTaskName": true,
            "isBuildCommand": true,
            "isWatching": false,
            "args": ["build"]
        },
        ...
    ]
    ...
}
```

That should assemble the full command to run as `npm run build`.

[EDIT] Although I do admit that if you don't have `supporessTaskName` that it should build the command correctly. Let me look into here and get back to you. In the meantime, try with the suppressed task name and see if that works.

**Comment by guardrex** [comment]
@GerjanOnline Yes, I just confirmed that I did set it up to run the command based on the `args` only. Therefore, you need to reconfigure your tasks to look just like this ...

``` json
{
    "taskName": "build",
    "suppressTaskName": true,
    "isBuildCommand": true,
    "isWatching": false,
    "args": ["build"]
},
```

The reason I did that was that I wanted to allow for you to name the tasks with `taskName` to anything you want ... that doesn't become part of the actual command that is run.

**Comment by GerjanOnline** [comment]
![image](https://cloud.githubusercontent.com/assets/1153419/13235439/2ec97996-d9be-11e5-8f85-6ce8de8f95f3.png)

Do I have the right config now? Something still not right... ;) CTRL + SHIFT + B stil works

**Comment by guardrex** [comment]
mmmm ... Try this for the args to see if it will write out and execute the full command.

``` json
"args" : ["npm", "run", "build"],
```

Some of my own problems are coming from the fact that I've only been using VS Code for two weeks!

**Comment by guardrex** [comment]
I sort of wonder, too, if you are going to have to change the `npm` to `cmd` and the top-level `args` from `["run"]` to `["/c"]` ... the way I show in the sample file.

Are you on Windows OS?

**Comment by guardrex** [comment]
So if you're on Windows, the file would look like this ...

``` json
{
    "version": "0.1.0",
    "command": "cmd"
    "isShellCommand": true,
    "showOutput": "always",
    "args": ["/c"],
    "tasks": [
        {
            "taskName": "build",
            "suppressTaskName": true,
            "isBuildCommand": true,
            "isWatching": false,
            "args": ["npm", "run", "build"]
        },
        ...
    ]
    ...
}
```

**Comment by GerjanOnline** [comment]
Yes that one works from the status bar but now the `Run Task` /  `CTRL+SHIFT+B` is broken! Its doing: `npm run npm run build`

**Comment by GerjanOnline** [comment]
Ah yes, I was also thinking in that direction, i need to switch to `cmd` as main command...

**Comment by guardrex** [comment]
Yes ... go with `cmd` there and `/c` as the arg.

**Comment by GerjanOnline** [comment]
Works! Thanks @GuardRex 

For completeness, my `tasks.json`:

``` javascript
{
  "version": "0.1.0",
  "command": "cmd",
  "isShellCommand": true,
  "showOutput": "always",
  "args": ["/c"],
  "tasks": [ 
    { 
        "taskName": "Build",
        "suppressTaskName": true,
        "echoCommand": true,
        "isBuildCommand": true,
        "isWatching": false,
        "args" : ["npm", "run", "build"]
    },
    { 
        "taskName": "Start-Dev",
        "suppressTaskName": true,
        "isBuildCommand": false,
        "isWatching": true,     
        "args" : ["npm", "run", "start-dev"]
    },
    { 
        "taskName": "Start-Prod",
        "suppressTaskName": true,
        "isBuildCommand": false,
        "isWatching": true,
        "args" : ["npm", "run", "start-prod"]   
    }
  ]
}
```

**Comment by guardrex** [comment]
@GerjanOnline Sweet! :smile: Before two weeks ago, I never even used VS Code. The extension is even newer. Sorry it took me so long, but I'm glad it had a happy ending.

I'll add some additional commentary on this to the ReadMe on my next update. All I say right now is that the format of the file must match, but I think I need to explain what the top-level `command` and `args` is doing relative to the `args` in the individual tasks. I'll explain it all out in my next update.

**Comment by Alphapage** [comment]
Hello,

I didn't have such a problem, but I wanted to use same command for all platforms with nodejs.
I only had to modify the getTaskArray function to make this working:

``` js
function getTasksArray() {
    try {
        var taskFilePath = path.join(vscode.workspace.rootPath, '.vscode', 'tasks.json');
        var rawTaskFileContents = fs.readFileSync(taskFilePath, 'utf8');
        var taskFileContents = rawTaskFileContents.replace(/((\/\/|\/\/\/)(.*)(\r\n|\r|\n))|((\/\*)((\D|\d)+)(\*\/))/gi, "");
        var taskFileTasks = JSON.parse(taskFileContents);
        if (taskFileTasks) {
            var taskElement = taskFileTasks['tasks'];
                    if (taskFileTasks.command){
                var command=taskFileTasks.command;
                if (taskFileTasks.args){
                var args = taskFileTasks.args;
                taskElement.forEach(function(task){
                    if (task.args){
                        task.args.splice(0,0,command)
                        task.args.splice.apply(task.args,[1,0].concat(args))
                    }
                    else {
                    task.args.push(command)
                        task.args.push(args)
                    }
                })
                }
            }
            return taskElement;
        }
        else {
            return null;
        }
    }
    catch (e) {
        return null;
    }
}
```

I debugged in the vscode directly, so this is pure javascript but should work in the ts file.
This code add the top-level command and args to each tasks when necessary.
I hope this will help you to enhance your tool.

**Comment by guardrex** [comment]
@Alphapage 

> This code add the top-level command and args to each tasks when necessary.

Yes ... its true that the F1 experience will execute the command along with whatever is in the task arguments and that your suggestion would make the extension behavior match the F1 experience exactly; however, the way I have it now where **only** the task argument list gets executed when the status bar button is clicked provides the greatest flexibility.

I mean that let's say someone were to want the top-level command and task arguments executed with the F1 experience but not for clicking the buttons? It seems rare ... I don't think many would want that. I'm just worried that by doing what you suggest that it would be locking in that top-level command everywhere when the user might not want it locked in like that (but still have it work for the F1 experience).

I sort of feel like a good enhancement here would be to have another switch on each task just for the status bar tasks that tells it whether or not to add the command to that task or not, which would provide very fine control.

Something like the following with logic to pickup and apply the `addTopLevelCommand` bool. In this example, there is an `npm` command in the top-level argument list. The first task will add the top-level command (and the `npm` command) to the task arguments, so `child_process` would be executing `cmd /c npm run build`. The second task will ignore the top-level command and `child_process` will only execute `npm run build` when clicked with nothing else added.

``` json
{
  "version": "0.1.0",
  "command": "cmd",
  "isShellCommand": true,
  "showOutput": "always",
  "args": ["/c", "npm"],
  "tasks": [ 
    { 
        "taskName": "Command in top-level command",
        "suppressTaskName": true,
        "echoCommand": true,
        "isBuildCommand": true,
        "args" : ["run", "build"],
        "addTopLevelCommand": true
    },
    { 
        "taskName": "Command in task arguments",
        "suppressTaskName": true,
        "isBuildCommand": true,  
        "args" : ["npm", "run", "build"],
        "addTopLevelCommand": false
    }
  ]
}
```

That all sounds good, but there is one lingering problem: The F1 experience built-in to VS Code will always execute the top-level command with the task arguments. That makes matching them difficult as @GerjanOnline and I were discussing. The F1 experience above will result in `cmd /c npm run buld` and `cmd /c npm npm run build`, so the second one will break.

**Comment by Alphapage** [comment]
I created a feature request:
https://github.com/Microsoft/vscode/issues/4475

I'm not agree with you because I think you have to follow the way vscode tasks.json is working for now, even if this is a way causing problems.
You just have to tell in the readme, that you don't care about command and args in top level.
This is not a problem for me because it shows the limitation of tasks.json.

And you can even implement your custom guardrextasks.json with the schema I propose in https://github.com/Microsoft/vscode/issues/4475.

**Comment by guardrex** [comment]
> you have to follow the way vscode tasks.json is working for now

Well, I don't totally disagree in general; however, Microsoft isn't Apple! :smile:, so I kind of can do whatever I want.

What I like about the way I have it now is that the Node `child_process` module will execute **exactly** what you specify at the task-level ... no more ... no less. Because it ignores the command at the top-level, you have the ultimately flexibility the way I set it up.

If you want all of your tasks to have a high level top command, simply add it to all of the tasks. If I set it up as you suggest without the switch that I was discussing, then it would force everyone into the VS Code-way of using the command on every task. I really, really think that's a bad idea here (but see my next comment ...).

> it shows the limitation of tasks.json

I agree with you 100% on that :+1:.  I'm not sure why it was setup with such a limited design as it currently has. Let's see if they respond to your feature request. I'll leave this open a bit longer for further discussion if they respond to your issue at Microsoft/vscode#4475.

**Comment by guardrex** [comment]
@Alphapage I pinged them on this, and I still plan to address the problem with the extension as soon as they fix it properly in VS Code. It looks like it slipped their April iteration. I'm hoping they'll get to it in May.

**Comment by guardrex** [comment]
Just released v0.2.0 that fixes this issue.

You can add and use a `command` property on individual tasks. If you do so, the global `command` and `args` are not applied.

If you leave your command and arguments in the `args` property and don't specify a task `command`, then the global `command` and `args` will be used.

```json
{
    "version": "0.1.0",
    "command": "cmd",
    "isShellCommand": true,
    "showOutput": "silent",
    "args": [
        "/c"
    ],
    "test": "blah blah blah",
    "tasks": [
        {
            "taskName": "restore that draws in the global command and args",
            "suppressTaskName": true,
            "args" : ["dotnet", "restore"],
            "showOutput": "always",
            "problemMatcher": "$msCompile"
        },
        {
            "taskName": "restore that doesn't draw in the global command and args",
            "suppressTaskName": true,
            "command": "dotnet",
            "args" : ["restore"],
            "showOutput": "always",
            "problemMatcher": "$msCompile"
        },
```
The first restore command will run ...
```
cmd /c dotnet restore
```
The second restore command will run ...
```
dotnet restore
```
